tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,271 @@
+import numpy as np
+from numpy.lib.recfunctions import unstructured_to_structured
+
+def naive_downsample(events: np.ndarray, sensor_size: tuple, target_size: tuple):
+    """Downsample the classic "naive" Tonic way. Multiply x/y values by a spatial_factor 
+    obtained by dividing sensor size by the target size.
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        sensor_size (tuple): a 3-tuple of x,y,p for sensor_size.
+        target_size (tuple): a 2-tuple of x,y denoting new down-sampled size for events to be
+                             re-scaled to (new_width, new_height).
+    Returns:
+        the downsampled input events.
+    """
+    
+    assert "x" and "y" in events.dtype.names
+    
+    events = events.copy()
+    
+    spatial_factor = np.asarray(target_size) / sensor_size[:-1]
+
+    events["x"] = events["x"] * spatial_factor[1]
+    events["y"] = events["y"] * spatial_factor[0]
+
+    return events
+
+def time_bin_numpy(events: np.ndarray, time_bin_interval: float):
+    """Temporally downsample the events into discrete time bins as stipulated by time_bin_intervals.
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        time_bin_interval (float): time bin size for events e.g. every 0.5 ms: time_bin_interval = 0.5.
+    Returns:
+        the input events with rewritten timestamps.
+    """
+    
+    events = events.copy()
+    reciprocal_interval = 1 / time_bin_interval
+    events["t"] = np.round(events["t"] * reciprocal_interval, 0) / reciprocal_interval
+    
+    return events
+
+def feedback_inhibition(events: np.ndarray, target_size: tuple, dt: float, buffer_layer_parameters: dict, inhibitory_layer_parameters: dict,
+                        buffer_postsynaptic_strength: float = 0.005, inhibitory_postsynaptic_strength: float = 4.0, delay: int = 1):
+    """Feedback inhibition to eliminate high-frequency noise inspired by locust brains.
+    Parameters:
```

Comment:
I think there needs to be a white space before 'Parameters' and 'Returns' in order to be rendered correctly in documentation

Reply:
Done